// react-photo-album/dist/scroll/index.js
"use client";
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { useMemo, useState, useCallback, cloneElement, useLayoutEffect as useLayoutEffect$1, useEffect, useRef, Children, isValidElement } from "react";
const observers = /* @__PURE__ */ new Map();
let rootId = 0;
const rootIds = /* @__PURE__ */ new WeakMap();
function getRootId(root) {
  if (!root) return 0;
  if (rootIds.has(root)) return rootIds.get(root);
  rootIds.set(root, ++rootId);
  return rootId;
}
function getObserverId(root, rootMargin) {
  return `${getRootId(root)}_${rootMargin}`;
}
function createObserver(root, rootMargin) {
  const id = getObserverId(root, rootMargin);
  let instance = observers.get(id);
  if (!instance) {
    const listeners = /* @__PURE__ */ new Map();
    const observer = new IntersectionObserver(
      (entries) => {
        entries.forEach((entry) => {
          listeners.get(entry.target)?.forEach((callback) => {
            callback(entry);
          });
        });
      },
      { root, rootMargin }
    );
    instance = { id, observer, listeners };
    observers.set(id, instance);
  }
  return instance;
}
function useIntersectionObserver(rootMargin, scrollContainer) {
  const root = scrollContainer?.() ?? null;
  return useMemo(() => {
    const cleanup = [];
    const observe = (target, callback) => {
      const { id, observer, listeners } = createObserver(root, rootMargin);
      let callbacks = listeners.get(target);
      if (!callbacks) {
        callbacks = /* @__PURE__ */ new Set();
        listeners.set(target, callbacks);
        observer.observe(target);
      }
      if (callbacks.has(callback)) {
        return;
      }
      callbacks.add(callback);
      cleanup.push(() => {
        callbacks.delete(callback);
        if (callbacks.size === 0) {
          listeners.delete(target);
          observer.unobserve(target);
        }
        if (listeners.size === 0) {
          observer.disconnect();
          observers.delete(id);
        }
      });
    };
    const unobserve = () => {
      cleanup.forEach((callback) => callback());
      cleanup.splice(0, cleanup.length);
    };
    return { observe, unobserve };
  }, [root, rootMargin]);
}
function Offscreen({ rootMargin, scrollContainer, children }) {
  const [placeholder, setPlaceholder] = useState();
  const { observe, unobserve } = useIntersectionObserver(rootMargin, scrollContainer);
  const ref = useCallback(
    (node) => {
      unobserve();
      if (node) {
        observe(node, ({ isIntersecting }) => {
          if (isIntersecting) {
            setPlaceholder(void 0);
            return;
          }
          const { width, height } = node.getBoundingClientRect();
          const { marginTop, marginRight, marginBottom, marginLeft } = getComputedStyle(node);
          setPlaceholder(
            (prev) => prev || {
              aspectRatio: `${width} / ${height}`,
              margin: `${marginTop} ${marginRight} ${marginBottom} ${marginLeft}`
            }
          );
        });
      }
    },
    [observe, unobserve]
  );
  return placeholder ? jsx("div", { ref, style: { width: "100%", ...placeholder } }) : cloneElement(children, { ref });
}
const useLayoutEffect = typeof window !== "undefined" ? useLayoutEffect$1 : useEffect;
function useEventCallback(fn) {
  const ref = useRef(fn);
  useLayoutEffect(() => {
    ref.current = fn;
  });
  return useCallback((...args) => ref.current?.(...args), []);
}
var Status = ((Status2) => {
  Status2[Status2["IDLE"] = 0] = "IDLE";
  Status2[Status2["LOADING"] = 1] = "LOADING";
  Status2[Status2["ERROR"] = 2] = "ERROR";
  Status2[Status2["FINISHED"] = 3] = "FINISHED";
  return Status2;
})(Status || {});
function InfiniteScroll({
  photos: initialPhotos,
  onClick,
  fetch,
  retries = 0,
  singleton,
  error,
  loading,
  finished,
  children,
  scrollContainer,
  fetchRootMargin = "800px",
  offscreenRootMargin = "2000px"
}) {
  const [status, setStatus] = useState(0);
  const [photos, setPhotos] = useState(() => initialPhotos ? [initialPhotos] : []);
  const { observe, unobserve } = useIntersectionObserver(fetchRootMargin, scrollContainer);
  const fetching = useRef(false);
  const fetchWithRetry = useEventCallback((index) => {
    let attempt = 1;
    const execute = async () => {
      try {
        return await fetch(index);
      } catch (err) {
        if (attempt > retries) {
          throw err;
        }
        await new Promise((resolve) => {
          setTimeout(resolve, 1e3 * 2 ** (attempt - 1));
        });
        attempt += 1;
        return execute();
      }
    };
    return execute();
  });
  const handleFetch = useCallback(async () => {
    if (fetching.current) return;
    const updateStatus = (newStatus) => {
      fetching.current = newStatus === 1;
      setStatus(newStatus);
    };
    updateStatus(1);
    try {
      const batch = await fetchWithRetry(photos.length);
      if (batch === null) {
        updateStatus(3);
        return;
      }
      setPhotos((prev) => [...prev, batch]);
      updateStatus(0);
    } catch (_) {
      updateStatus(2);
    }
  }, [photos.length, fetchWithRetry]);
  const sentinelRef = useCallback(
    (node) => {
      unobserve();
      if (node) {
        observe(node, ({ isIntersecting }) => {
          if (isIntersecting) {
            handleFetch();
          }
        });
      }
    },
    [observe, unobserve, handleFetch]
  );
  const photosArray = photos.flatMap((batch) => batch);
  const handleClick = onClick ? (offset) => ({ photo, index, event }) => {
    onClick({ photos: photosArray, index: offset + index, photo, event });
  } : void 0;
  let batchOffset = 0;
  return jsxs(Fragment, { children: [
    singleton ? cloneElement(children, {
      photos: photosArray,
      onClick: handleClick?.(0),
      render: {
        ...children.props.render,
        track: ({ children: trackChildren, ...rest }) => jsx("div", { ...rest, children: Children.map(
          trackChildren,
          (child, index) => isValidElement(child) && jsx(Offscreen, { rootMargin: offscreenRootMargin, scrollContainer, children: child }, index)
        ) })
      }
    }) : photos.map((batch, index) => {
      const offset = batchOffset;
      batchOffset += batch.length;
      return jsx(Offscreen, { rootMargin: offscreenRootMargin, scrollContainer, children: cloneElement(children, {
        photos: batch,
        onClick: handleClick?.(offset)
      }) }, index);
    }),
    status === 2 && error,
    status === 1 && loading,
    status === 3 && finished,
    jsx("span", { ref: sentinelRef })
  ] });
}
export {
  InfiniteScroll as default
};
